File: 2020/15.py
NUMS = [0, 3, 1, 6, 7, 5]


def spoken_number(nums, idx):
    mem = {v: [idx] for idx, v in enumerate(nums)}

    turn = len(nums)
    prev = nums[-1]
    while turn < idx:
        if prev not in mem or len(mem[prev]) == 1:
            n = 0
        else:
            history = mem[prev]
            n = history[1] - history[0]
        if n not in mem:
            mem[n] = [turn]
        else:
            mem[n] = (mem[n] + [turn])[-2:]
        prev = n
        turn += 1
    return prev


# part 1
print(spoken_number(NUMS, 2020))

# part 2
print(spoken_number(NUMS, 30000000))

File: 2020/15.cpp
#include <iostream>
#include <vector>

using namespace std;

const uint32_t nums[] = {0, 3, 1, 6, 7, 5};
const size_t nums_size = sizeof(nums) / sizeof(uint32_t);


uint32_t spoken_number(size_t N) {
    vector<uint32_t> mem(N, 0);
    for (size_t i = 0; i < nums_size-1; ++i) {
        mem[nums[i]] = i + 1;
    }

    uint32_t prev = nums[nums_size-1];
    for (size_t turn = nums_size; turn < N; ++turn) {
        uint32_t next = 0;
        if (mem[prev] != 0) {
            next = turn - mem[prev];
        }
        mem[prev] = turn;
        prev = next;
    }
    return prev;
}

int main(int argc, char **argv) {
    auto start = chrono::steady_clock::now();
    auto p1 = spoken_number(2020);
    auto p2 = spoken_number(30000000);
    auto end = chrono::steady_clock::now();

    cout << "part 1: " << p1 << endl;
    cout << "part 2: " << p2 << endl;

    cout << "Runtime: " << chrono::duration_cast<chrono::microseconds>(end-start).count() << " us" << endl;
}
